src/charm.py

```diff
@@ -3,61 +3,77 @@
+
+
+class NotLeaderError(Exception):
+    def __init__(self):
+        super().__init__('not leader')
+        self.status = WaitingStatus('Not leader')
 
 
 class GitLabK8sCharm(CharmBase):
     state = StoredState()
+    on = GitLabK8sCharmEvents()
 
     def __init__(self, framework, key):
         super().__init__(framework, key)
 
-        for event in (self.on.start,
-                      self.on.upgrade_charm,
-                      self.on.config_changed,
-                      self.on.mysql_relation_changed):
-            self.framework.observe(event, self.on_start)
-        self.framework.observe(self.on.website_relation_joined, self)
-
-        self.website = HTTPInterfaceProvides(self, 'website')
-        self.mysql = MySQLInterfaceRequires(self, 'mysql')
+        self.website = HTTPServer(self, 'website')
+        self.mysql = MySQLServer(self, 'mysql')
         self.gitlab_image = OCIImageResource(self, 'gitlab_image')
 
+        self.state.is_started = getattr(self.state, 'is_started', False)
+        self.state.has_website = getattr(self.state, 'has_website', False)
 
-    def on_install(self, event):
-        self.state.is_started = False
+        self.framework.observe(self.on.start, self.check_ready)  # TODO: change to install
+        self.framework.observe(self.on.leader_elected, self.check_ready)
+        self.framework.observe(self.on.upgrade_charm, self.check_ready)
+        self.framework.observe(self.on.config_changed, self.check_ready)
+        self.framework.observe(self.mysql.on.database_available, self.check_ready)
+        self.framework.observe(self.mysql.on.database_changed, self.check_ready)
+        self.framework.observe(self.on.configure_pod, self.configure_pod)
+        self.framework.observe(self.website.on.new_client, self.serve_website)
 
-    def on_start(self, event):
-        unit = self.framework.model.unit
-        if not self.gitlab_image.fetch():
-            unit.status = BlockedStatus('Missing or invalid image resource')
-            return
-        if not self.mysql.is_joined:
-            unit.status = BlockedStatus('Missing database')
-        if not self.mysql.is_single:
-            unit.status = BlockedStatus('Too many databases')
-        if not self.mysql.is_ready:
-            unit.status = WaitingStatus('Waiting for database')
-            return
-        if not self.framework.model.unit.is_leader():
-            unit.status = WaitingStatus('Not leader')
-            return
-        unit.status = MaintenanceStatus('Configuring container')
-        self.framework.model.pod.set_spec({
+    def check_ready(self, event):
+        try:
+            self.mysql.database()
+            self.gitlab_image.fetch()
+            if not self.model.unit.is_leader():
+                raise NotLeaderError()
+        except (MySQLServerError, FetchError, NotLeaderError) as e:
+            self.model.unit.status = e.status
+        else:
+            self.on.configure_pod.emit()
```

Comment:
Why do this via an event, instead of just calling self.configure_pod? (in which case you can also pass in the Database that you grabbed earlier, and whatever information you get from gitlab_image.fetch()

The fact that fetch() doesn't really return anything, but updates the attributes of gitlab_image, would it be better to call it something like ensure() ?